### ISSUE TITLE
Issue with Program Freezing When Using DP

### ISSUE THREAD

**Issue by miliable** [AISIGSJTU/Siren-Plus#1]
Hello, author,I encountered the following issue when running the code with the command `"python main.py --dataset=fMNIST --k=10 --C=1.0 --E=5 --T=40 --B=64 --train --model_num=1 --mal --gar=siren --attack_type=label_flipping --dp --dp_mul=1.0"`. Could you please help me understand why the program freezes when I use DP (Differential Privacy), while it runs normally without DP? I would greatly appreciate your advice on what might be causing this issue and look forward to your response. Thank you very much.
![image](https://github.com/user-attachments/assets/1cf4112b-c3ec-4a40-b363-65a045cce2ff)

**Comment by MarkGHX** [comment]
Thank you for raising this issue. Could you please provide more detailed information about your system (e.g., OS version) and experimental environment (including packages and their versions, which can be obtained via ```conda list``` or ```pip list```)? This information will help us reproduce the problem you encountered.

**Comment by miliable** [comment]
In my computer(Windows),conda list:
# Name                    Version                   Build  Channel
absl-py                   0.15.0                   pypi_0    pypi
astunparse                1.6.3                    pypi_0    pypi
ca-certificates           2024.8.30            h56e8100_0    conda-forge
cachetools                5.5.0                    pypi_0    pypi
certifi                   2024.8.30                pypi_0    pypi
charset-normalizer        3.4.0                    pypi_0    pypi
cudatoolkit               11.0.3              hd336c7a_13    conda-forge
cudnn                     8.0.5.39             hfe7f257_1    conda-forge
debugpy                   1.8.7                    pypi_0    pypi
dm-tree                   0.1.8                    pypi_0    pypi
flatbuffers               1.12                     pypi_0    pypi
gast                      0.3.3                    pypi_0    pypi
google-auth               2.35.0                   pypi_0    pypi
google-auth-oauthlib      0.4.6                    pypi_0    pypi
google-pasta              0.2.0                    pypi_0    pypi
grpcio                    1.32.0                   pypi_0    pypi
h5py                      2.10.0                   pypi_0    pypi
idna                      3.10                     pypi_0    pypi
importlib-metadata        8.5.0                    pypi_0    pypi
joblib                    1.4.2                    pypi_0    pypi
keras                     2.3.1                    pypi_0    pypi
keras-applications        1.0.8                    pypi_0    pypi
keras-preprocessing       1.1.2                    pypi_0    pypi
libsqlite                 3.46.1               h2466b09_0    conda-forge
markdown                  3.7                      pypi_0    pypi
markupsafe                2.1.5                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
numpy                     1.19.5                   pypi_0    pypi
oauthlib                  3.2.2                    pypi_0    pypi
openssl                   3.3.2                h2466b09_0    conda-forge
opt-einsum                3.3.0                    pypi_0    pypi
pandas                    1.4.1                    pypi_0    pypi
pip                       24.2               pyh8b19718_1    conda-forge
protobuf                  3.20.3                   pypi_0    pypi
pyasn1                    0.6.1                    pypi_0    pypi
pyasn1-modules            0.4.1                    pypi_0    pypi
python                    3.8.12          h900ac77_2_cpython    conda-forge
python-dateutil           2.9.0.post0              pypi_0    pypi
pytz                      2024.2                   pypi_0    pypi
pyyaml                    6.0.2                    pypi_0    pypi
requests                  2.32.3                   pypi_0    pypi
requests-oauthlib         2.0.0                    pypi_0    pypi
rsa                       4.9                      pypi_0    pypi
scikit-learn              1.0.1                    pypi_0    pypi
scipy                     1.10.1                   pypi_0    pypi
setuptools                75.1.0             pyhd8ed1ab_0    conda-forge
six                       1.15.0                   pypi_0    pypi
sqlite                    3.46.1               h2466b09_0    conda-forge
tensorboard               2.11.2                   pypi_0    pypi
tensorboard-data-server   0.6.1                    pypi_0    pypi
tensorboard-plugin-wit    1.8.1                    pypi_0    pypi
tensorflow                2.4.1                    pypi_0    pypi
tensorflow-estimator      2.4.0                    pypi_0    pypi
tensorflow-gpu            2.4.1                    pypi_0    pypi
tensorflow-privacy        0.5.1                    pypi_0    pypi
termcolor                 1.1.0                    pypi_0    pypi
threadpoolctl             3.5.0                    pypi_0    pypi
typing-extensions         3.7.4.3                  pypi_0    pypi
ucrt                      10.0.22621.0         h57928b3_1    conda-forge
urllib3                   2.2.3                    pypi_0    pypi
vc                        14.3                ha32ba9b_22    conda-forge
vc14_runtime              14.40.33810         hcc2c482_22    conda-forge
vs2015_runtime            14.40.33810         h3bf8584_22    conda-forge
werkzeug                  3.0.4                    pypi_0    pypi
wheel                     0.44.0             pyhd8ed1ab_0    conda-forge
wrapt                     1.12.1                   pypi_0    pypi
zipp                      3.20.2                   pypi_0    pypi

My OS version in Linux is Ubuntu 20.04.4 LTS
conda list:
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
absl-py                   0.15.0                   pypi_0    pypi
astunparse                1.6.3                    pypi_0    pypi
bzip2                     1.0.8                h4bc722e_7    conda-forge
ca-certificates           2024.8.30            hbcca054_0    conda-forge
cachetools                5.5.0                    pypi_0    pypi
certifi                   2024.8.30          pyhd8ed1ab_0    conda-forge
charset-normalizer        3.4.0                    pypi_0    pypi
cudatoolkit               11.0.3              h7761cd4_13    conda-forge
cudnn                     8.0.5.39             ha5ca753_1    conda-forge
dm-tree                   0.1.8                    pypi_0    pypi
flatbuffers               1.12                     pypi_0    pypi
gast                      0.3.3                    pypi_0    pypi
google-auth               2.35.0                   pypi_0    pypi
google-auth-oauthlib      0.4.6                    pypi_0    pypi
google-pasta              0.2.0                    pypi_0    pypi
grpcio                    1.32.0                   pypi_0    pypi
h5py                      2.10.0                   pypi_0    pypi
idna                      3.10                     pypi_0    pypi
importlib-metadata        8.5.0                    pypi_0    pypi
joblib                    1.4.2                    pypi_0    pypi
keras                     2.3.1                    pypi_0    pypi
keras-applications        1.0.8                    pypi_0    pypi
keras-preprocessing       1.1.2                    pypi_0    pypi
ld_impl_linux-64          2.43                 h712a8e2_1    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc                    14.2.0               h77fa898_1    conda-forge
libgcc-ng                 14.2.0               h69a702a_1    conda-forge
libgomp                   14.2.0               h77fa898_1    conda-forge
libnsl                    2.0.1                hd590300_0    conda-forge
libsqlite                 3.46.0               hde9e2c9_0    conda-forge
libstdcxx                 14.2.0               hc0a3c3a_1    conda-forge
libstdcxx-ng              14.2.0               h4852527_1    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libzlib                   1.2.13               h4ab18f5_6    conda-forge
markdown                  3.7                      pypi_0    pypi
markupsafe                2.1.5                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
ncurses                   6.5                  he02047a_1    conda-forge
numpy                     1.23.5                   pypi_0    pypi
oauthlib                  3.2.2                    pypi_0    pypi
openssl                   1.1.1w               hd590300_0    conda-forge
opt-einsum                3.3.0                    pypi_0    pypi
pandas                    1.4.1                    pypi_0    pypi
pip                       24.2               pyh8b19718_1    conda-forge
protobuf                  3.20.3                   pypi_0    pypi
pyasn1                    0.6.1                    pypi_0    pypi
pyasn1-modules            0.4.1                    pypi_0    pypi
python                    3.8.15          h257c98d_0_cpython    conda-forge
python-dateutil           2.9.0.post0              pypi_0    pypi
pytz                      2024.2                   pypi_0    pypi
pyyaml                    6.0.2                    pypi_0    pypi
readline                  8.2                  h8228510_1    conda-forge
requests                  2.32.3                   pypi_0    pypi
requests-oauthlib         2.0.0                    pypi_0    pypi
rsa                       4.9                      pypi_0    pypi
scikit-learn              1.0.1                    pypi_0    pypi
scipy                     1.10.1                   pypi_0    pypi
setuptools                75.1.0             pyhd8ed1ab_0    conda-forge
six                       1.15.0                   pypi_0    pypi
sqlite                    3.46.0               h6d4b2fc_0    conda-forge
tensorboard               2.11.2                   pypi_0    pypi
tensorboard-data-server   0.6.1                    pypi_0    pypi
tensorboard-plugin-wit    1.8.1                    pypi_0    pypi
tensorflow-estimator      2.4.0                    pypi_0    pypi
tensorflow-gpu            2.4.0                    pypi_0    pypi
tensorflow-privacy        0.5.1                    pypi_0    pypi
termcolor                 1.1.0                    pypi_0    pypi
threadpoolctl             3.5.0                    pypi_0    pypi
tk                        8.6.13          noxft_h4845f30_101    conda-forge
typing-extensions         3.7.4.3                  pypi_0    pypi
urllib3                   2.2.3                    pypi_0    pypi
werkzeug                  3.0.4                    pypi_0    pypi
wheel                     0.44.0             pyhd8ed1ab_0    conda-forge
wrapt                     1.12.1                   pypi_0    pypi
xz                        5.2.6                h166bdaf_0    conda-forge
zipp                      3.20.2                   pypi_0    pypi

They all experience program freezing due to using DP, but it runs normally without using DP.

**Comment by MarkGHX** [comment]
Thank you for your reply. Additionally, could you let me know how long you waited for the program execution? Based on our previous tests, when LDP is enabled, the client execution may take several extra minutes compared to the version with LDP disabled.

**Comment by MarkGHX** [comment]
Besides, I noticed that some of the packages you are using differ slightly from our recommended versions. If convenient, could you try the following commands to create a new virtual environment and test the code again while we work on reproducing the issue?

```
conda create -n siren-plus python=3.8.12
conda activate siren-plus
conda install tensorflow-gpu==2.4.1
pip install tensorflow-privacy==0.5.1
conda install keras==2.3.1 scikit-learn pandas
```

I just tested our code with this configuration, and it worked well. Thanks for your time.

**Comment by miliable** [comment]
Thank you very much for your response. I followed your method and successfully solved the problem.

**Comment by MarkGHX** [comment]
Thank you for your reply! We will update our environment configuration to make deployment easier. Since the problem has been resolved, I will close this issue. If you encounter any further problems, please feel free to open a new issue. Thank you for your time!

**Comment by MarkGHX** [comment]
Issue fixed.
